### ISSUE TITLE
New default avatar

### ISSUE THREAD

**Issue by driesvints** [laravelio/laravel.io#688]
Figure our a way to show a grayed out Laravel logo as a default avatar instead of the current default GitHub one.

**Comment by driesvints** [comment]
Gonna re-open this to keep our target to also figure out a way to replace the default GitHub avatar. Maybe we can sync through GitHub web hooks to figure out if a user is using a default GitHub avatar or not.

**Comment by driesvints** [comment]
To be clear: I mean the coloured block ones in this screenshot should also be the greyed out Laravel.io icon.

<img width="854" alt="Screenshot 2022-01-08 at 16 22 24" src="https://user-images.githubusercontent.com/594614/148649775-a3e0744a-d65a-457d-ae52-7e63ae9f25c3.png">

**Comment by driesvints** [comment]
Additionally, we should filter out empty avatars on the homepage so we have a full wall of photo's.

**Comment by driesvints** [comment]
Might also want to remove unavatar in total because we seem to be running into rate limiting issues. Not sure what the best way forward is here.

**Comment by ProjektGopher** [comment]
- get user info from api `https://api.github.com/users/USERNAME`
- on 200, temporarily download image `$response->avatar_url`
- if image !png, return
- if image !420x420, return
- if image has > 2 colours, return (more resource intensive)

Not sure if I'll have enough time to sit and implement this, but thought I'd leave these notes regardless

**Comment by driesvints** [comment]
oh nice @ProjektGopher. Thanks a lot for digging into that! Definitely useful for follow up.

**Comment by imacrayon** [comment]
> Might also want to remove unavatar

We could store a user’s GitHub avatar URL in the DB when they register. Then we could check if the URL needs to be updated each time a user logs in or through a scheduled job.

**Comment by driesvints** [comment]
@imacrayon yeah I was first thinking about GitHub webhooks but that might be a bit too complex. A middleware on logged in routes that checks the updated_at timestamp of a logged in user and rechecks the avatar might work.

**Comment by tauseefsshah** [comment]
Hi @driesvints 

So I found something for the coloured github default avatar thing.

We can get the current user avatar from the `https://api.github.com/users/{USERNAME}` api and then we can verify if it's the default [github identicon](https://github.blog/news-insights/company-news/identicons/) by comparing with the `https://github.com/identicons/{USERNAME}.png` (This url has the default github identicon png for the user)

I found this package [image-comparator-laravel](https://github.com/sapientpro/image-comparator-laravel) and it has 100% accuracy for this (I tried on 5 samples).

Let me know if this package and approach looks good. I'll take it up :)

**Comment by driesvints** [comment]
Hah that's an elaborate way to go about it @tauseefsshah! Do you think a PR could fulfil all the requirements listed above?

**Comment by tauseefsshah** [comment]
@driesvints Following is my understanding of the pending requirements:

1. If a user has a default github avatar then show grayed out laravel.io icon
2. Filter out empty avatars and only have ones with an actual photo

Following is my solution:

1. Create columns `has_default_avatar`, `avatar_synced_at` and `github_profile_updated_at` in `users` table
2. Write a command that runs every 6hrs to run a job to update the `has_default_avatar` field if the last sync is >24hrs (The job will only run if the `github_profile_updated_at` is not same as `updated_at` returned by github api)

Let me know if I am missing on any requirement or should change the approach.

**Comment by tauseefsshah** [comment]
@driesvints Following is my understanding of the pending requirements:

1. If a user has a default github avatar then show grayed out laravel.io icon
2. Filter out empty avatars and only have ones with an actual photo

Following is my solution:

1. Create columns `has_default_avatar`, `avatar_synced_at` and `github_profile_updated_at` in `users` table
2. Write a command that runs every 6hrs to run a job to update the `has_default_avatar` field if the last sync is >24hrs (The job will only run if the `github_profile_updated_at` is not same as `updated_at` returned by github api)

Let me know if I am missing on any requirement or should change the approach.

**Comment by driesvints** [comment]
Yes that's correct. Although I'm not sure we'd want to add that many columns. Tbh I'd like to see a more simpler solution. Maybe we should let this issue be for what it is for now.

**Comment by Nathan-Cowin** [comment]
@driesvints I noticed on the [github identicon](https://github.blog/news-insights/company-news/identicons/) blog post tauseefsshah found it said "for anyone without a Gravatar" - therefore everyone with a Github & Gravatar will have a profile picture
![image](https://github.com/user-attachments/assets/205da377-59ba-4495-9a6b-c97e228c1ab7)

They have a rest API so ...
- if user found - user avatar
- if user not found - default github icon

**Notes**
Api Limits
- Unauthenticated requests are limited to 100 per hour.
- Authenticated requests are limited to 1000 per hour.

Even though it says "for anyone without a Gravatar" I quickly made some curl calls and I believe you can have github profile picture but gravatar api returns no profile - So you can have "false-negatives" but shouldn't have "false-positives" I hope that makes sense.

In other words it would remove all the default github icons, but you might lose some users avatars. If this is a tradeoff you're willing to live with I'm happy to code it up

**Comment by driesvints** [comment]
Maybe I should just implement profile image uploads to keep things even simpler 😂

**Comment by driesvints** [comment]
I managed to solve the unavatar issue pretty easily by loading from github directly: https://github.com/laravelio/laravel.io/commit/bdef8d0b81af20d01bf0662f5bb2a3f2748d1d9a
